src/test/java/com/google/devtools/build/lib/starlark/StarlarkRuleImplementationFunctionsTest.java

```diff
@@ -2555,6 +2556,152 @@ public void testLazyArgsObjectImmutability() throws Exception {
+        "a.append(actions.args().add(Label('//bar')))",
+        "a.append(actions.args().add('-flag', Label('//bar')))",
+        "a.append(actions.args().add('-flag', Label('//bar'), format = '_%s_'))",
+        "a.append(actions.args().add_all(['foo', Label('//bar')]))",
+        "a.append(actions.args().add_all(depset([Label('//foo'), Label('//bar')])))",
+        "ruleContext.actions.run(",
+        "  inputs = depset(ruleContext.files.srcs),",
+        "  outputs = ruleContext.files.srcs,",
+        "  arguments = a,",
+        "  executable = ruleContext.files.tools[0],",
+        "  toolchain = None",
+        ")");
+    SpawnAction action =
+        (SpawnAction)
+            Iterables.getOnlyElement(
+                ruleContext.getRuleContext().getAnalysisEnvironment().getRegisteredActions());
+    assertThat(action.getArguments())
+        .containsExactly(
+            "foo/t.exe",
+            "//bar:bar",
+            "-flag",
+            "//bar:bar",
+            "-flag",
+            "_//bar:bar_",
+            "foo",
+            "//bar:bar",
+            "//foo:foo",
+            "//bar:bar")
+        .inOrder();
+  }
+
+  @Test
+  public void testArgsCanonicalRepoLabel() throws Exception {
+    StarlarkRuleContext ruleContext = createRuleContext("//foo:foo");
+    setRuleContext(ruleContext);
+    ev.exec(
+        "actions = ruleContext.actions",
+        "a = []",
+        "a.append(actions.args().add(Label('@@repo~//:foo')))",
+        "a.append(actions.args().add('-flag', Label('@@repo~//:foo')))",
+        "a.append(actions.args().add('-flag', Label('@@repo~//:foo'), format = '_%s_'))",
+        "a.append(actions.args().add_all(['foo', Label('@@repo~//:foo')]))",
+        "a.append(actions.args().add_all(depset([Label('@@other_repo~//:foo'), Label('@@repo~//:foo')])))",
+        "ruleContext.actions.run(",
+        "  inputs = depset(ruleContext.files.srcs),",
+        "  outputs = ruleContext.files.srcs,",
+        "  arguments = a,",
+        "  executable = ruleContext.files.tools[0],",
+        "  toolchain = None",
+        ")");
+    SpawnAction action =
+        (SpawnAction)
+            Iterables.getOnlyElement(
+                ruleContext.getRuleContext().getAnalysisEnvironment().getRegisteredActions());
+    assertThat(action.getArguments())
+        .containsExactly(
+            "foo/t.exe",
+            "@@repo~//:foo",
+            "-flag",
+            "@@repo~//:foo",
+            "-flag",
+            "_@@repo~//:foo_",
+            "foo",
+            "@@repo~//:foo",
+            "@@other_repo~//:foo",
+            "@@repo~//:foo")
+        .inOrder();
+  }
+
+  @Test
+  public void testArgsApparentRepoLabel() throws Exception {
+    StarlarkRuleContext ruleContext = createRuleContext("//foo:foo");
+    setRuleContext(ruleContext);
+    ev.exec(
+        "actions = ruleContext.actions",
+        "a = []",
+        // bazel_tools is a well-known Bazel module whose canonical repo name is bazel_tools.
```

Comment:
this test fails internally during import, because the internal test setup doesn't use built-in modules. Could you please explicitly add a MODULE.bazel file here? (and an explicit `bazel_dep` on some dep, etc etc)

Reply:
I added an empty `MODULE.bazel` file as no `bazel_dep` is needed for this test. I'm not sure whether I have to call `invalidatePackages()` though, which I think is needed when making changes to `WORKSPACE`. Feel free to add it during import if it turns out to be necessary (or ask me to add it).

Reply:
the problem is that `bazel_tools` doesn't even exist. What I meant was that you'd add a `bazel_dep` on like `foo@1.0` or something and test `@foo` labels are returned.

Reply:
I failed to get this to work: If I use a module such as `rules_java`, I get an error about it being built-in. If I use a non-built-in module such as `stardoc`, it isn't found (likely since the test doesn't access the BCR). If I use a `local_path_override` to add a module `foo`, I get:
```
Unrecoverable error while evaluating node 'foo' (requested by nodes 'ConfiguredTargetKey{label=//foo:foo, config=BuildConfigurationKey[7a8260b91962440adc5c50c32c1d6004d1f32e31cea089bab75639b2331de7c7]}')
java.lang.RuntimeException: 
	at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:553)
	at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:414)
	at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1403)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
Caused by: java.lang.IllegalStateException: Unexpected Exception type from PackageValue for 'foo'' with error: ErrorInfo{exception=java.io.IOException: to fix, run
	bazel fetch //...
External repository @@foo~ not found and fetching repositories is disabled., cycles=[], isCatastrophic=false, isDirectlyTransient=false, isTransitivelyTransient=true}
	at com.google.devtools.build.lib.skyframe.packages.AbstractPackageLoader.exceptionFromErrorInfo(AbstractPackageLoader.java:408)
	at com.google.devtools.build.lib.skyframe.packages.AbstractPackageLoader.loadPackagesInternal(AbstractPackageLoader.java:388)
	at com.google.devtools.build.lib.skyframe.packages.AbstractPackageLoader.loadPackages(AbstractPackageLoader.java:367)
	at com.google.devtools.build.lib.skyframe.packages.AbstractPackageLoader.loadPackage(AbstractPackageLoader.java:353)
	at com.google.devtools.build.lib.testutil.BazelPackageLoadingListenerForTesting.sanityCheckBazelPackageLoader(BazelPackageLoadingListenerForTesting.java:85)
	at com.google.devtools.build.lib.testutil.BazelPackageLoadingListenerForTesting.onLoadingCompleteAndSuccessful(BazelPackageLoadingListenerForTesting.java:61)
	at com.google.devtools.build.lib.packages.PackageFactory.afterDoneLoadingPackage(PackageFactory.java:331)
	at com.google.devtools.build.lib.skyframe.PackageFunction.compute(PackageFunction.java:548)
	at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:464)
	... 7 more
Caused by: java.io.IOException: to fix, run
	bazel fetch //...
External repository @@foo~ not found and fetching repositories is disabled.
	at com.google.devtools.build.lib.rules.repository.RepositoryDelegatorFunction.compute(RepositoryDelegatorFunction.java:223)
	... 8 more
```

Do you see a pattern that could work here?

Reply:
could you do something like this? https://cs.opensource.google/bazel/bazel/+/master:src/test/java/com/google/devtools/build/lib/skyframe/toolchains/RegisteredToolchainsFunctionTest.java;drc=09bf5472ae15237471ee298c194836ab0067d7c5;l=377